open_spiel/games/battleship.cc

```diff
@@ -0,0 +1,849 @@
+// Copyright 2020 DeepMind Technologies Ltd. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.#include "open_spiel/games/battleship.h"
+
+#include "open_spiel/games/battleship.h"
+
+#include "open_spiel/abseil-cpp/absl/strings/ascii.h"
+#include "open_spiel/abseil-cpp/absl/strings/numbers.h"
+#include "open_spiel/abseil-cpp/absl/strings/str_split.h"
+#include "open_spiel/abseil-cpp/absl/strings/string_view.h"
+#include "open_spiel/abseil-cpp/absl/strings/strip.h"
+
+namespace open_spiel {
+namespace battleship {
+
+const double kFloatTolerance = 1e-9;
+
+const Player Player1 = Player{0};
+const Player Player2 = Player{1};
+
+BattleshipState::BattleshipState(
+    const std::shared_ptr<const BattleshipGame> bs_game)
+    : State(bs_game), bs_game_(bs_game) {}
+
+Player BattleshipState::CurrentPlayer() const {
+  const BattleshipConfiguration& conf = bs_game_->configuration;
+
+  // The players place the ships on the board in turns, starting from Player 1.
+  //
+  // NOTE: It is important whether or not the players place all their ships at
+  // once or not for correlated equilibria purposes. This is because in
+  // correlated equilibria, the recommender can stop issuing recommendations
+  // after a player deviates from a recommended *action*.
+  if (!AllShipsPlaced()) {
+    // In this case, if an even number (possibly 0) of ships have been placed,
+    // then it is Player 1's turn to act next. Else, it is Player 2's.
+    if (NumShipsPlaced() % 2 == 0) {
+      return Player1;
+    } else {
+      return Player2;
+    }
+  } else {
```

Comment:
Avoid deep nesting of code when all previous cases have been handled (returned from function). It makes the code more readable. I usually stick to max 3 or 4 levels of indentation, if more is needed then I put it in another function.

Reply:
I think I prefer keeping the `else`s explicitly there, unless other maintainers also agree that the code as written is not readable enough. I understand that you *can* early-return from functions in C, but I think that `else` actually aids readability, because it signals that the two cases are completely disjoint and you can reason about the two branches independently. Also, it makes code more robust, just in case something changes and all of a sudden things are not early-returning any more.

At any rate I don't feel super strongly about this so I'm OK to change things around if there if folks feel strongly.

Reply:
I prefer having the else in these cases for the same reasons but I also don't feel particularly strongly about it.

Reply:
OK, then I'm going to resolve this for now. Feel free to unresolve if anything changes.